src/GraphQL.SystemTextJson/ExecutionResultJsonConverter.cs

```diff
@@ -29,8 +30,64 @@ private void WriteData(Utf8JsonWriter writer, ExecutionResult result)
                 return;
             }
 
-            writer.WritePropertyName("data");
-            JsonSerializer.Serialize(writer, data);
+            WriteDataRecursively(writer, "data", data);
+        }
+
+        private void WriteDataRecursively(Utf8JsonWriter writer, string name, object data)
+        {
+            if (name != null)
+            {
+                if (data == null)
+                {
+                    writer.WriteNull(name);
+                    return;
+                }
+
+                if (data is string s)
+                {
+                    writer.WriteString(name, s);
+                    return;
+                }
+
+                if (data is bool b)
+                {
+                    writer.WriteBoolean(name, b);
+                    return;
+                }
+
+                writer.WritePropertyName(name);
+            }
+
+            if (data is Dictionary<string, object> dictionary)
+            {
+                writer.WriteStartObject();
+
+                foreach (var kvp in dictionary)
+                    WriteDataRecursively(writer, kvp.Key, kvp.Value);
+
+                writer.WriteEndObject();
+            }
+            else if (data is List<object> list)
+            {
+                writer.WriteStartArray();
+
+                foreach (object item in list)
+                {
+                    if (item is null)
+                        writer.WriteNullValue();
+                    else if (item is string s)
+                        writer.WriteStringValue(s);
+                    else
+                        WriteDataRecursively(writer, null, item);
+                }
+
+                writer.WriteEndArray();
+            }
+            else
+            {
+                // Need to avoid this call by all means! The question remains open - why this API so expensive?
```

Comment:
@davidfowl could you explain please?  Another question that interests me is whether the serialization options for ` writer.WriteString(name, s);` and other low level `WriteXXX` APIs apply? For example `PropertyNamingPolicy`.

Reply:
Yea the docs are a bit funky saying you shouldn't pass down the options as it could cause a stackoverflow or something. So you have to recreate them at every level it seemed to say. I avoided that by just not adding any tests that test for it :P

Pretty sure the docs say to use the static JsonSerializer.Serialize in examples, so it's kinda weird this makes such a difference.

Reply:
> Another question that interests me is whether the serialization options for writer.WriteString(name, s); and other low level WriteXXX APIs apply? For example PropertyNamingPolicy.

It looks like the answer is **no**. `PropertyNamingPolicy` is called by `JsonPropertyInfo.DeterminePropertyName` when building `JsonClassInfo` from `Type`.

Therefore the adventure continues.

Reply:
In fact, some of the options apply, for example `WriteIndented`.